### DISCUSSION TITLE
LiveView in a POS Extension

### DISCUSSION THREAD

**Discussion by kenichi** [ericdude4/shopifex#62] [Q&A]
Hi and thanks for a great library @ericdude4 et al. I'm building a [POS Links](https://shopify.dev/docs/apps/pos/links) extension and just can not get LiveView to connect without hitting [this message](https://github.com/phoenixframework/phoenix_live_view/blob/v0.17.14/lib/phoenix_live_view/channel.ex#L845) which I suspect has to do with sessions, cookies, native-app webviews, CSP headers...

Instead I'm writing JS, using appbridge session tokens, and `:shopify_api` pipeline, but would sure love to use LiveView.

Do you know what I might be doing wrong or of a way to make this work?

**Comment by ericdude4** [comment]
Hey! Very interesting application for LiveView :)

I believe Shopify might have some very restrictive session requirements which may be blocking LiveView's ability to maintain a session. What is your configuration in your `Endpoint` module?
```
@session_options [
                  ...
                ]
```

**Reply by kenichi** [reply]
Thanks - I hope it can work 🤞🏽 

Here's that attribute with redacted/example data:

```
  @session_options [
    store: :cookie,
    key: "_example_app_key",
    signing_salt: "...",
    encryption_salt: "..."
  ]
```
Note that this is happening through both the POS iOS app and the App URL/admin, but direct liveviews connect and work fine.

```
[info] CONNECTED TO Phoenix.LiveView.Socket in 22µs
  Transport: :websocket
  Serializer: Phoenix.Socket.V2.JSONSerializer
  Parameters: %{"_csrf_token" => "...", "_mounts" => "0", "_track_static" => %{"0" => "https://example.com/assets/ap
p.css", "1" => "https://example.com/assets/app.js"}, "vsn" => "2.0.0"}
[debug] LiveView session was misconfigured or the user token is outdated.

1) Ensure your session configuration in your endpoint is in a module attribute:
...
```

Here's the CSP response header that the `:shopify_embedded` pipeline is setting:

```
{"content-security-policy",
   "frame-ancestors https://admin.shopify.com https://example.myshopify.com;"}
```

**Reply by ericdude4** [reply]
@kenichi I'm not sure exactly what the right fix is, but I suspect the cause is due to Shopify cookie restrictions in that iFrame which you are being rendered in. In your `@session_options`, I think `store: :cookie` and `store: :ets` will both not work because they both rely on the presence of a cookie.

Perhaps you can try something like this when you connect to your socket:

```javascript
let liveSocket = new LiveSocket("/live", Socket, {params: {shop_token: <%= Shopifex.Plug.session_token(conn) %>}})
liveSocket.connect()
```

Maybe you can use this `shop_token` in the `mount()` callback to fetch the shop and put it in the session securely.

There are also some other steps you should take to remove cookie reliance from your pipeline:
https://byteflip.de/blog/cookieless-setup

excited to hear how this goes for you!

**Comment by rgraff** [comment]
You need to create a second LiveView socket that doesn't have use sessions, then copy the session data you need from the connection to the socket when the LiveView mounts. See my readme changes in this PR:

https://github.com/ericdude4/shopifex/pull/79/files

**Reply by kenichi** [reply]
@rgraff interesting, i'll have to look back at how i made it work, but i don't use two LV sockets, and i do auth on the socket conn, by implementing my  own "Phoenix.LiveView.Socket". i'll look at your PR as well - thanks for commenting on this old discussion!

**Reply by ericdude4** [reply]
@kenichi Would love to see what you ended up implementing in your custom Socket!

**Reply by rgraff** [reply]
@kenichi Have you done any LiveView polaris components? I've started to create a few just to get a sense of how time consuming that would be

**Reply by kenichi** [reply]
ok i perhaps misunderstood, because i don't use two LV sockets when in shopify admin/POS, but my app does expose one socket for shopify, and another "normal" LV socket. only one is used at a time. i wanted the shopify socket to be authenticated by the same session token, and i get the `appState` value in the process. i'm curious - @rgraff is the LV socket you use authed at all or does it allow arbitrary connections? and wow that forum post would have been super helpful to me 2 years ago! 😆 i've not made any polaris components, i was angling for POS extensions, which are now deprecated for react. sadly, i've been building new things for this app in that and using phoenix as an API.

NOTE: i'm still on phoenix 1.7.14 and LV 0.18.18 for this app.

in endpoint.ex:

```elixir
   # "normal" LV socket
   socket "/live", Phoenix.LiveView.Socket, websocket: [connect_info: [session: @session_options]]

  # shopify admin/pos LV socket
  socket "/shopify/live", ExampleWeb.Shopify.Socket
```

in shopify/socket.ex:

```elixir
  def connect(params, %Phoenix.Socket{} = socket, connect_info) do
    with {:ok, token} <- Map.fetch(params, "sessionToken"),
         {:ok, _claims} <- Shopifex.Guardian.decode_and_verify(token),
         {:ok, shop, _claims} <- Guardian.resource_from_token(Shopifex.Guardian, token),
         app_state <- Map.get(params, "appState", %{}),
         connect_info <- build_connect_info(connect_info, app_state, shop) do
      {:ok, put_private(socket, :connect_info, connect_info)}
    else
      error ->
        Logger.error(inspect(error))
        debug_token(params)
        :error
    end
  end

  defp put_private(socket, key, value), do: put_in(socket.private[key], value)

  defp build_connect_info(connect_info, app_state, shop) do
    connect_info
    |> Map.put(:app_state, app_state)
    |> Map.put(:shop, shop)
  end
```

in JS:

```javascript
  if (isShopifyEmbedded()) {
    ...
        window.app = window.CreateAppBridge({
          apiKey: getShopifyClientId(),
          host: usp.get("host"),
          forceRedirect: true,
        });

        await liveSocketSetup();
    ...
  } else {
    liveSocketSetup();
  }
```

```javascript
function getCSRFToken() {
  return document
    .querySelector("meta[name='csrf-token']")
    .getAttribute("content");
}

async function params() {
  let params = { _csrf_token: getCSRFToken() };

  if (isShopifyEmbedded()) {
    const token = await window.getSessionToken(window.app);
    params["sessionToken"] = token;

    const appState = await getAppState();
    params["appState"] = appState;
  }

  return params;
}

function path() {
  return isShopifyEmbedded() ? "/shopify/live" : "/live";
}

class InMemoryStorage {
  constructor() { this.storage = {} }
  getItem(keyName) { return this.storage[keyName] || null }
  removeItem(keyName) { delete this.storage[keyName] }
  setItem(keyName, keyValue) { this.storage[keyName] = keyValue }
}

async function options() {
  let options = {
    localStorage: new InMemoryStorage(),
    params: await params(),
    sessionStorage: new InMemoryStorage(),
  };

  return options;
}

export async function liveSocketSetup() {
  const liveSocket = new LiveSocket(path(), Socket, await options());
  if (isShopifyEmbedded()) {
    const phxSocket = liveSocket.getSocket();
    phxSocket.onError((_) => {
      liveSocket.disconnect();
      toastPOSError("Connection error! Please reload...");
    });
  }

  liveSocket.connect();
  window.liveSocket = liveSocket;

  return;
}

**Reply by rgraff** [reply]
That's a very similar approach to what I'm doing, except I'm using the `session` option provided by `live_session` to copy the token into the socket. My solution relies on `shopify_session` plug to authenticate the shop.
